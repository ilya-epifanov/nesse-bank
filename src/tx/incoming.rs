use crate::{account::AccountId, Money};
use serde::{Deserialize, Serialize};
use thiserror::Error;

use super::TxId;

#[derive(Error, Debug)]
pub enum IncomingTxError {
    #[error("error parsing amount: {0}")]
    ParseAmount(#[from] rust_decimal::Error),
    #[error("negative amount")]
    NegativeAmount,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
#[repr(C)]
pub enum IncomingTxDetails {
    Deposit(Money),
    Withdrawal(Money),
    Dispute,
    Resolve,
    Chargeback,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
#[repr(C)]
pub struct IncomingTx {
    pub id: TxId,
    pub account: AccountId,
    pub details: IncomingTxDetails,
}

impl IncomingTxDetails {
    pub fn balance_effect(&self) -> Option<Money> {
        match self {
            IncomingTxDetails::Deposit(amount) => Some(*amount),
            IncomingTxDetails::Withdrawal(amount) => Some(-amount),
            _ => None,
        }
    }
}

impl IncomingTx {
    pub fn deposit(
        id: impl Into<TxId>,
        account: impl Into<AccountId>,
        amount: &str,
    ) -> Result<Self, IncomingTxError> {
        let amount = Money::from_str_exact(amount)?.ensure_non_negative()?;
        Ok(Self {
            id: id.into(),
            account: account.into(),
            details: IncomingTxDetails::Deposit(amount),
        })
    }

    pub fn withdrawal(
        id: impl Into<TxId>,
        account: impl Into<AccountId>,
        amount: &str,
    ) -> Result<Self, IncomingTxError> {
        let amount = Money::from_str_exact(amount)?.ensure_non_negative()?;
        Ok(Self {
            id: id.into(),
            account: account.into(),
            details: IncomingTxDetails::Withdrawal(amount),
        })
    }

    pub fn dispute(id: impl Into<TxId>, account: impl Into<AccountId>) -> Self {
        Self {
            id: id.into(),
            account: account.into(),
            details: IncomingTxDetails::Dispute,
        }
    }

    pub fn resolve(id: impl Into<TxId>, account: impl Into<AccountId>) -> Self {
        Self {
            id: id.into(),
            account: account.into(),
            details: IncomingTxDetails::Resolve,
        }
    }

    pub fn chargeback(id: impl Into<TxId>, account: impl Into<AccountId>) -> Self {
        Self {
            id: id.into(),
            account: account.into(),
            details: IncomingTxDetails::Chargeback,
        }
    }
}

trait MoneyExt: Sized {
    fn ensure_non_negative(self) -> Result<Self, IncomingTxError>;
}

impl MoneyExt for Money {
    fn ensure_non_negative(mut self) -> Result<Self, IncomingTxError> {
        if self.is_sign_negative() && !self.is_zero() {
            Err(IncomingTxError::NegativeAmount)
        } else {
            self.set_sign_positive(true);
            Ok(self)
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::Money;

    use super::MoneyExt;

    #[test]
    fn money_ensure_non_negative_makes_zero_positive() {
        assert_eq!(
            Money::from_str_exact("-0.0")
                .unwrap()
                .ensure_non_negative()
                .unwrap(),
            Money::from_str_exact("0.0").unwrap()
        );
    }

    #[test]
    fn money_ensure_non_negative_fails_on_negative() {
        assert_eq!(
            Money::from_str_exact("-1.0")
                .unwrap()
                .ensure_non_negative()
                .ok(),
            None
        );
    }

    #[test]
    fn money_ensure_non_negative_leaves_positive_untouched() {
        assert_eq!(
            Money::from_str_exact("1.0")
                .unwrap()
                .ensure_non_negative()
                .unwrap(),
            Money::from_str_exact("1.0").unwrap()
        );
    }
}
