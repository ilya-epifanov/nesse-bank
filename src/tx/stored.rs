use std::fmt::Debug;

use kv::Raw;
use serde::{Deserialize, Serialize};

use super::incoming::IncomingTx;

#[derive(Copy, Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
#[repr(C)]
pub struct TxDetails {
    pub original_tx: IncomingTx,
    pub state: TxState,
}

impl TxDetails {
    pub fn with_state(self, state: TxState) -> Self {
        Self {
            original_tx: self.original_tx,
            state,
        }
    }
}
const TX_DETAILS_SIZE: usize = std::mem::size_of::<TxDetails>();

impl From<Raw> for TxDetails {
    fn from(raw: Raw) -> Self {
        debug_assert_eq!(raw.len(), TX_DETAILS_SIZE);

        // MaybeUninit is still quirky with arrays on stable
        // Also, zeroing out will be optimized away in --release
        let mut bytes = [0; TX_DETAILS_SIZE];
        bytes.copy_from_slice(&raw);

        // SAFETY
        // safe because TxDetails is Copy
        unsafe { std::mem::transmute(bytes) }
    }
}

impl From<TxDetails> for Raw {
    fn from(tx: TxDetails) -> Self {
        // SAFETY
        // safe because TxDetails is Copy
        let v: [u8; TX_DETAILS_SIZE] = unsafe { std::mem::transmute(tx) };
        Raw::from(&v)
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "kebab-case")]
#[repr(C)]
pub enum TxState {
    Complete,
    UnderDispute,
    Resolved,
    ChargedBack,
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use kv::Raw;

    use crate::tx::incoming::IncomingTx;

    use super::TxDetails;

    #[test]
    fn tx_details_to_raw_and_back() {
        let original = TxDetails {
            original_tx: IncomingTx::deposit(123, 456, "789.1112").unwrap(),
            state: super::TxState::Complete,
        };

        let raw: Raw = original.clone().into();
        assert_debug_snapshot!(&raw);

        let recovered: TxDetails = raw.into();
        assert_eq!(recovered, original);
    }
}
