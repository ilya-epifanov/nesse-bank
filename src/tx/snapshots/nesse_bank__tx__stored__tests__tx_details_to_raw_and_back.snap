---
source: src/tx/stored.rs
assertion_line: 74
expression: "&raw"
---
[
    123,
    0,
    0,
    0,
    200,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    4,
    0,
    0,
    0,
    0,
    0,
    168,
    104,
    120,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
]
